File: tally.go
// How to run:
// go run tally.go
package main

import (
	"fmt"
	"sync"
	"time"
)

const COUNT = 50_000_000

func main() {
	var i int64

	positives := []int64{}
	for i = 1; i <= COUNT; i++ {
		positives = append(positives, i)
	}

	negatives := []int64{}
	for i = -1; i >= -COUNT; i-- {
		negatives = append(negatives, i)
	}

	var total int64 = 0
	tally := func(wg *sync.WaitGroup, numbers []int64) {
		defer wg.Done()

		fmt.Println("  inner job: tallying", len(numbers), "numbers...")
		startTime := time.Now()
		for _, n := range numbers {
			total += n
		}
		totalTime := time.Since(startTime)
		fmt.Println("    took", totalTime.Seconds(), "seconds")
	}

	wg := sync.WaitGroup{}
	fmt.Println("outer job: tallying positives and negatives...")
	startTime := time.Now()
	wg.Add(1)
	go tally(&wg, positives)
	wg.Add(1)
	go tally(&wg, negatives)
	wg.Wait()
	totalTime := time.Since(startTime)
	fmt.Println("outer job took", totalTime.Seconds(), "seconds total")
	fmt.Println("total sum:", total)
}

File: tally.mjs
// How to run:
// node tally.mjs
const COUNT = 50_000_000;

// fill `positives` array with numbers from 1 to 50,000,000
const positives = [];
for (let i = 1; i <= COUNT; i++) {
  positives.push(i);
}

// fill `negatives` array with numbers from -1 to -50,000,000
const negatives = [];
for (let i = -1; i >= -COUNT; i--) {
  negatives.push(i);
}

let total = 0;
// adds all numbers in `numbers` into `total`
async function tally(numbers) {
  console.log("  inner job: tallying", numbers.length, "numbers...");
  const startTime = Date.now();
  for (const n of numbers) { total += n }
  const totalTime = Date.now() - startTime;
  console.log("    took", totalTime / 1000, "seconds");
}


console.log("outer job: tallying positives and negatives...");
const startTime = Date.now();
await Promise.all([tally(positives), tally(negatives)]);
const totalTime = Date.now() - startTime;
console.log("outer job took", totalTime / 1000, "seconds total");

console.log("total sum:", total);
